src/Neo/SmartContract/Native/NativeContract.cs

```diff
@@ -138,49 +188,93 @@ protected NativeContract()
-            this.Hash = Helper.GetContractHash(UInt160.Zero, 0, Name);
-            this.Manifest = new ContractManifest
+            nef.CheckSum = NefFile.ComputeChecksum(nef);
+
+            // Compose manifest
+            ContractManifest manifest = new()
             {
                 Name = Name,
                 Groups = Array.Empty<ContractGroup>(),
                 SupportedStandards = Array.Empty<string>(),
                 Abi = new ContractAbi()
                 {
-                    Events = Array.Empty<ContractEventDescriptor>(),
-                    Methods = descriptors.Select(p => p.Descriptor).ToArray()
+                    Events = eventsDescriptors
+                        .Where(u => u.ActiveIn is null || settings.IsHardforkEnabled(u.ActiveIn.Value, index))
+                        .Select(p => p.Descriptor).ToArray(),
+                    Methods = allowedMethods.Methods.Values
+                        .Select(p => p.Descriptor).ToArray()
                 },
                 Permissions = new[] { ContractPermission.DefaultPermission },
                 Trusts = WildcardContainer<ContractPermissionDescriptor>.Create(),
                 Extra = null
             };
-            contractsList.Add(this);
-            contractsDictionary.Add(Hash, this);
+
+            OnManifestCompose(manifest);
+
+            // Return ContractState
+            return new ContractState
+            {
+                Id = Id,
+                Nef = nef,
+                Hash = Hash,
+                Manifest = manifest
+            };
         }
 
+        protected virtual void OnManifestCompose(ContractManifest manifest) { }
+
         /// <summary>
         /// It is the initialize block
         /// </summary>
         /// <param name="settings">The <see cref="ProtocolSettings"/> where the HardForks are configured.</param>
         /// <param name="index">Block index</param>
+        /// <param name="hardfork">Active hardfork</param>
         /// <returns>True if the native contract must be initialized</returns>
-        internal bool IsInitializeBlock(ProtocolSettings settings, uint index)
+        internal bool IsInitializeBlock(ProtocolSettings settings, uint index, out Hardfork? hardfork)
         {
-            if (ActiveIn is null) return index == 0;
+            // If is not configured, the Genesis is the a initialized block
+            if (index == 0 && ActiveIn is null)
+            {
+                hardfork = null;
+                return true;
+            }
 
-            if (!settings.Hardforks.TryGetValue(ActiveIn.Value, out var activeIn))
+            // If is in the hardfork height, return true
+            foreach (Hardfork hf in listenHardforks)
             {
-                return false;
+                if (settings.Hardforks.TryGetValue(hf, out var activeIn) && activeIn == index)
+                {
+                    hardfork = hf;
+                    return true;
+                }
```

Comment:
```C#
            // If is in the hardfork height, return true
            foreach (Hardfork hf in listenHardforks)
            {
                // if the hardfork is specificed in the configuration
                if (settings.Hardforks.TryGetValue(hf, out var activeIn))
                {
                    // if the hardfork is active in the block
                    if (activeIn == index)
                    {
                        hardfork = hf;
                        return true;
                    }
                }
                else if (!settings.Hardforks.ContainsKey(hf))
                {
                    // Get the enum value for the current hardfork
                    int currentHfValue = Convert.ToInt32(hf);

                    // Check if there are any hardforks configured with a higher enum value than the current one
                    if (settings.Hardforks.Any(val => Convert.ToInt32(val) > currentHfValue))
                    {
                        // If there are subsequent configured hardforks, activate this one immediately
                        hardfork = hf;
                        return true;
                    }
                    // If not, do nothing as the current hardfork will not be activated
                }
            }
```

Reply:
Thanks for your review. Please, check it again :)

Reply:
Shargon, please take a look at the code above, the logic here is if A and B both in the code while only B in the config, then A is enabled since block 0. but if A is in config while B is not, then B should be disabled by default.

Reply:
I think we should validate and avoid these situations. This logic should be faster as possible.

Reply:
Why we need this case (https://github.com/neo-project/neo/pull/2886):

- disable new hardforks if they're not specified in the config (A: 100500 in the config, but we know there is also B, but it's not configured, so it's disabled)

We should validate it, and B should be mandatory, if you don't want to, just put 9999999999999999.

Reply:
In that code if A,B and C is not present, all was enabled at 0. right?

Reply:
@Liaojinghui @roman-khimov @AnnaShaleva 

I think we have complicated the configuration of the hardforks too much, if you have a node and you simply forget to update the configuration, you will have a different configuration without being warned, in my opinion, I prefer an exception, it's better that you always have to specify ALL the hardforks, or none, and If it is none, all start with 0 (private net), but for any configuration you must set ALL.

Reply:
The rationale here is rather simple, consider a network starting at HF C from the genesis. No one will bother specifying HFs in the config. The HF D comes and we need to configure it. What I'd like to see in the config is
```
Hardforks:
    D: X
```
What will be required if not for this behavior is
```
Hardforks:
  A: 0
  B: 0
  C: 0
  D: X
```
Meh.

And HF configuration mismatch can happen in a lot of ways leading to the same instability.